diff -ur glib-2.54.3-orig/configure.ac glib-2.54.3/configure.ac
--- glib-2.54.3-orig/configure.ac	2018-01-08 20:19:22.000000000 +0000
+++ glib-2.54.3/configure.ac	2019-03-26 22:13:54.428930213 +0000
@@ -981,7 +981,7 @@
                                                           [int qclass = C_IN;])],
                                          [AC_MSG_RESULT([yes])
                                           NAMESER_COMPAT_INCLUDE="#include <arpa/nameser_compat.h>"],
-                                         [AC_MSG_ERROR([could not compile test program either way])])])])
+                                         [AC_MSG_RESULT([no])])])])
   AC_SUBST(NAMESER_COMPAT_INCLUDE)
 
   # We can't just use AC_CHECK_FUNC/AC_CHECK_LIB here. Bug 586150
@@ -1009,7 +1009,7 @@
                                         [res_query("test", 0, 0, (void *)0, 0);],
                                         [AC_MSG_RESULT([in -lbind])
                                          NETWORK_LIBS="-lbind $NETWORK_LIBS"],
-                                        [AC_MSG_ERROR(not found)])])
+                                        [AC_MSG_RESULT(no)])])
                LIBS="$save_libs"])
   AC_CHECK_FUNC(socket, :, AC_CHECK_LIB(socket, socket,
 				        [NETWORK_LIBS="-lsocket $NETWORK_LIBS"],
diff -ur glib-2.54.3-orig/gio/glocalfile.c glib-2.54.3/gio/glocalfile.c
--- glib-2.54.3-orig/gio/glocalfile.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/gio/glocalfile.c	2019-03-26 22:13:54.432263551 +0000
@@ -2800,7 +2800,7 @@
   const gchar *name;
   GDir *dir;
 
-#ifdef AT_FDCWD
+#if defined AT_FDCWD && !defined SWITCH
   {
     /* If this fails, we want to preserve the errno from fopendir() */
     DIR *dirp;
diff -ur glib-2.54.3-orig/gio/gnetworking.h.in glib-2.54.3/gio/gnetworking.h.in
--- glib-2.54.3-orig/gio/gnetworking.h.in	2017-07-14 00:03:38.000000000 +0100
+++ glib-2.54.3/gio/gnetworking.h.in	2019-03-26 22:13:54.432263551 +0000
@@ -38,17 +38,17 @@
 
 #include <sys/types.h>
 
+#include <sys/socket.h>
 #include <netdb.h>
 #include <netinet/in.h>
 #include <netinet/tcp.h>
-#include <resolv.h>
-#include <sys/socket.h>
-#include <sys/un.h>
+//#include <resolv.h>
+//#include <sys/un.h>
 #include <net/if.h>
 
 #include <arpa/inet.h>
-#include <arpa/nameser.h>
-@NAMESER_COMPAT_INCLUDE@
+//#include <arpa/nameser.h>
+//@NAMESER_COMPAT_INCLUDE@
 
 #ifndef T_SRV
 #define T_SRV 33
diff -ur glib-2.54.3-orig/gio/gsocketaddress.c glib-2.54.3/gio/gsocketaddress.c
--- glib-2.54.3-orig/gio/gsocketaddress.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/gio/gsocketaddress.c	2019-03-26 22:13:54.432263551 +0000
@@ -35,7 +35,7 @@
 #include "glibintl.h"
 #include "gioenumtypes.h"
 
-#ifdef G_OS_UNIX
+#if defined G_OS_UNIX && !defined(SWITCH)
 #include "gunixsocketaddress.h"
 #endif
 
@@ -265,7 +265,7 @@
       return sockaddr;
     }
 
-#ifdef G_OS_UNIX
+#if defined G_OS_UNIX && !defined(SWITCH)
   if (family == AF_UNIX)
     {
       struct sockaddr_un *addr = (struct sockaddr_un *) native;
diff -ur glib-2.54.3-orig/gio/gsocket.c glib-2.54.3/gio/gsocket.c
--- glib-2.54.3-orig/gio/gsocket.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/gio/gsocket.c	2019-03-26 22:13:54.435596889 +0000
@@ -48,7 +48,7 @@
 # include <sys/filio.h>
 #endif
 
-#ifdef G_OS_UNIX
+#if defined G_OS_UNIX && !defined(SWITCH)
 #include <sys/uio.h>
 #endif
 
@@ -2226,7 +2226,7 @@
   native_addr = g_inet_address_to_bytes (group);
   if (g_inet_address_get_family (group) == G_SOCKET_FAMILY_IPV4)
     {
-#ifdef HAVE_IP_MREQN
+#if defined HAVE_IP_MREQN && !defined(SWITCH)
       struct ip_mreqn mc_req;
 #else
       struct ip_mreq mc_req;
@@ -2235,7 +2235,7 @@
       memset (&mc_req, 0, sizeof (mc_req));
       memcpy (&mc_req.imr_multiaddr, native_addr, sizeof (struct in_addr));
 
-#ifdef HAVE_IP_MREQN
+#if defined HAVE_IP_MREQN && !defined(SWITCH)
       if (iface)
         mc_req.imr_ifindex = if_nametoindex (iface);
       else
@@ -2269,6 +2269,7 @@
     }
   else if (g_inet_address_get_family (group) == G_SOCKET_FAMILY_IPV6)
     {
+#ifndef SWITCH
       struct ipv6_mreq mc_req_ipv6;
 
       memset (&mc_req_ipv6, 0, sizeof (mc_req_ipv6));
@@ -2283,6 +2284,9 @@
       optname = join_group ? IPV6_JOIN_GROUP : IPV6_LEAVE_GROUP;
       result = setsockopt (socket->priv->fd, IPPROTO_IPV6, optname,
 			   &mc_req_ipv6, sizeof (mc_req_ipv6));
+#else
+      result = -1;
+#endif
     }
   else
     g_return_val_if_reached (FALSE);
@@ -3971,7 +3975,7 @@
   #endif
 }
 
-#ifndef G_OS_WIN32
+#if !defined G_OS_WIN32 && !defined SWITCH
 
 /* Unfortunately these have to be macros rather than inline functions due to
  * using alloca(). */
@@ -4328,7 +4332,7 @@
       vectors = &one_vector;
     }
 
-#ifndef G_OS_WIN32
+#if !defined G_OS_WIN32 && !defined SWITCH
   {
     GOutputMessage output_message;
     struct msghdr msg;
@@ -4379,7 +4383,7 @@
 
     return result;
   }
-#else
+#elif defined(G_OS_WIN32)
   {
     struct sockaddr_storage addr;
     guint addrlen;
@@ -4459,6 +4463,10 @@
 
     return bytes_sent;
   }
+#else
+ g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+                             _("Stubbed on switch"));
+  return -1;
 #endif
 }
 
@@ -4557,7 +4565,7 @@
   if (num_messages == 0)
     return 0;
 
-#if !defined (G_OS_WIN32) && defined (HAVE_SENDMMSG)
+#if !defined (G_OS_WIN32) && defined (HAVE_SENDMMSG) && !defined(SWITCH)
   {
     struct mmsghdr *msgvec;
     gint i, num_sent;
@@ -4793,7 +4801,7 @@
       vectors = &one_vector;
     }
 
-#ifndef G_OS_WIN32
+#if !defined G_OS_WIN32 && !defined SWITCH
   {
     GInputMessage input_message;
     struct msghdr msg;
@@ -4861,7 +4869,7 @@
 
     return result;
   }
-#else
+#elif defined G_OS_WIN32
   {
     struct sockaddr_storage addr;
     int addrlen;
@@ -4945,6 +4953,10 @@
 
     return bytes_received;
   }
+#else
+ g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+                             _("Stubbed on switch"));
+  return -1;
 #endif
 }
 
@@ -5063,7 +5075,7 @@
   if (num_messages == 0)
     return 0;
 
-#if !defined (G_OS_WIN32) && defined (HAVE_RECVMMSG)
+#if !defined (G_OS_WIN32) && defined (HAVE_RECVMMSG) && !defined(SWITCH)
   {
     struct mmsghdr *msgvec;
     guint i, num_received;
diff -ur glib-2.54.3-orig/gio/gthreadedresolver.c glib-2.54.3/gio/gthreadedresolver.c
--- glib-2.54.3-orig/gio/gthreadedresolver.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/gio/gthreadedresolver.c	2019-03-26 22:13:54.435596889 +0000
@@ -35,6 +35,7 @@
 #include "gsocketaddress.h"
 #include "gsrvtarget.h"
 
+#ifndef SWITCH
 
 G_DEFINE_TYPE (GThreadedResolver, g_threaded_resolver, G_TYPE_RESOLVER)
 
@@ -955,3 +956,5 @@
   addrinfo_hints.ai_socktype = SOCK_STREAM;
   addrinfo_hints.ai_protocol = IPPROTO_TCP;
 }
+
+#endif
\ No newline at end of file
diff -ur glib-2.54.3-orig/gio/gunixmounts.c glib-2.54.3/gio/gunixmounts.c
--- glib-2.54.3-orig/gio/gunixmounts.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/gio/gunixmounts.c	2019-03-26 22:13:54.435596889 +0000
@@ -837,7 +837,18 @@
 
   return return_list;
 }
+#elif defined(SWITCH)
+static char *
+get_mtab_monitor_file (void)
+{
+  return NULL;
+}
 
+static GList *
+_g_get_unix_mounts (void)
+{
+  return NULL;
+}
 /* Common code {{{2 */
 #else
 #error No _g_get_unix_mounts() implementation for system
@@ -1355,7 +1366,12 @@
 {
   return _g_get_unix_mounts ();
 }
-
+#elif defined(SWITCH)
+static GList *
+_g_get_unix_mount_points (void)
+{
+  return NULL;
+}
 /* Common code {{{2 */
 #else
 #error No g_get_mount_table() implementation for system
diff -ur glib-2.54.3-orig/gio/gunixsocketaddress.c glib-2.54.3/gio/gunixsocketaddress.c
--- glib-2.54.3-orig/gio/gunixsocketaddress.c	2017-07-14 00:03:39.000000000 +0100
+++ glib-2.54.3/gio/gunixsocketaddress.c	2019-03-26 22:13:54.438930228 +0000
@@ -28,7 +28,7 @@
 #include "glibintl.h"
 #include "gnetworking.h"
 
-
+#ifndef SWITCH
 /**
  * SECTION:gunixsocketaddress
  * @short_description: UNIX GSocketAddress
@@ -564,3 +564,4 @@
   return FALSE;
 #endif
 }
+#endif
\ No newline at end of file
diff -ur glib-2.54.3-orig/gio/Makefile.am glib-2.54.3/gio/Makefile.am
--- glib-2.54.3-orig/gio/Makefile.am	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/gio/Makefile.am	2019-03-26 22:13:54.438930228 +0000
@@ -793,7 +793,8 @@
 gio-2.0.lib: libgio-2.0.la gio.def
 	$(AM_V_GEN) lib.exe -machine:@LIB_EXE_MACHINE_FLAG@ -name:libgio-2.0-$(LT_CURRENT_MINUS_AGE).dll -def:$(builddir)/gio.def -out:$@
 
-bin_PROGRAMS = gio-querymodules glib-compile-schemas glib-compile-resources gsettings
+#bin_PROGRAMS = gio-querymodules glib-compile-schemas glib-compile-resources gsettings
+bin_PROGRAMS =
 
 glib_compile_resources_LDADD = libgio-2.0.la 		\
 	$(top_builddir)/gobject/libgobject-2.0.la	\
@@ -875,7 +876,7 @@
 # ------------------------------------------------------------------------
 # gdbus(1) tool
 
-bin_PROGRAMS += gdbus
+#bin_PROGRAMS += gdbus
 gdbus_SOURCES = gdbus-tool.c
 gdbus_LDADD = libgio-2.0.la 				\
 	$(top_builddir)/gobject/libgobject-2.0.la	\
@@ -887,7 +888,7 @@
 if !OS_COCOA
 # ------------------------------------------------------------------------
 # gapplication(1) tool
-bin_PROGRAMS += gapplication
+#bin_PROGRAMS += gapplication
 gapplication_SOURCES = gapplication-tool.c
 gapplication_LDADD = libgio-2.0.la 			\
 	$(top_builddir)/gobject/libgobject-2.0.la	\
@@ -908,7 +909,7 @@
 # ------------------------------------------------------------------------
 # gresource tool
 
-bin_PROGRAMS += gresource
+#bin_PROGRAMS += gresource
 gresource_SOURCES = gresource-tool.c
 gresource_CPPFLAGS = $(LIBELF_CFLAGS) $(AM_CPPFLAGS)
 gresource_LDADD = libgio-2.0.la				\
@@ -920,7 +921,7 @@
 # ------------------------------------------------------------------------
 # gio tool
 
-bin_PROGRAMS += gio
+#bin_PROGRAMS += gio
 gio_SOURCES = \
 	gio-tool.c					\
 	gio-tool.h					\
diff -ur glib-2.54.3-orig/glib/deprecated/gthread.h glib-2.54.3/glib/deprecated/gthread.h
--- glib-2.54.3-orig/glib/deprecated/gthread.h	2017-07-14 00:03:39.000000000 +0100
+++ glib-2.54.3/glib/deprecated/gthread.h	2019-03-26 22:13:54.438930228 +0000
@@ -123,7 +123,7 @@
 void     g_thread_foreach      (GFunc             thread_func,
                                 gpointer          user_data);
 
-#ifndef G_OS_WIN32
+#if !defined G_OS_WIN32 && !defined SWITCH
 #include <sys/types.h>
 #include <pthread.h>
 #endif
@@ -133,7 +133,7 @@
 typedef struct
 {
   GMutex *mutex;
-#ifndef G_OS_WIN32
+#if !defined G_OS_WIN32 && !defined SWITCH
   /* only for ABI compatibility reasons */
   pthread_mutex_t unused;
 #endif
@@ -162,7 +162,7 @@
 
   /* ABI compat only */
   union {
-#ifdef G_OS_WIN32
+#if defined G_OS_WIN32 || defined SWITCH
     void *owner;
 #else
     pthread_t owner;
diff -ur glib-2.54.3-orig/glib/gfileutils.c glib-2.54.3/glib/gfileutils.c
--- glib-2.54.3-orig/glib/gfileutils.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/glib/gfileutils.c	2019-03-26 22:13:54.438930228 +0000
@@ -417,7 +417,9 @@
   
   if ((test & G_FILE_TEST_IS_EXECUTABLE) && (access (filename, X_OK) == 0))
     {
+#ifndef SWITCH
       if (getuid () != 0)
+#endif
 	return TRUE;
 
       /* For root, on some POSIX systems, access (filename, X_OK)
@@ -430,10 +432,12 @@
 
   if (test & G_FILE_TEST_IS_SYMLINK)
     {
+#ifndef SWITCH
       struct stat s;
 
       if ((lstat (filename, &s) == 0) && S_ISLNK (s.st_mode))
         return TRUE;
+#endif
     }
   
   if (test & (G_FILE_TEST_IS_REGULAR |
diff -ur glib-2.54.3-orig/glib/giounix.c glib-2.54.3/glib/giounix.c
--- glib-2.54.3-orig/glib/giounix.c	2017-09-11 01:09:54.000000000 +0100
+++ glib-2.54.3/glib/giounix.c	2019-03-26 22:13:54.438930228 +0000
@@ -184,10 +184,13 @@
 {
   GIOUnixChannel *unix_channel = (GIOUnixChannel *)channel;
   gssize result;
-
+#ifdef SWITCH
+  if (count > 32767) /* At least according to the Debian manpage for read */
+    count = 32767;
+#else
   if (count > SSIZE_MAX) /* At least according to the Debian manpage for read */
     count = SSIZE_MAX;
-
+#endif
  retry:
   result = read (unix_channel->fd, buf, count);
 
diff -ur glib-2.54.3-orig/glib/glib-unix.c glib-2.54.3/glib/glib-unix.c
--- glib-2.54.3-orig/glib/glib-unix.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/glib/glib-unix.c	2019-04-04 20:22:43.474122263 +0100
@@ -84,6 +84,7 @@
                   int      flags,
                   GError **error)
 {
+#ifndef SWITCH
   int ecode;
 
   /* We only support FD_CLOEXEC */
@@ -127,6 +128,10 @@
       return g_unix_set_error_from_errno (error, saved_errno);
     }
   return TRUE;
+#else
+  *error = g_error_new(1, 1, "pipes don't work here");
+  return FALSE;
+#endif
 }
 
 /**
diff -ur glib-2.54.3-orig/glib/gmain.c glib-2.54.3/glib/gmain.c
--- glib-2.54.3-orig/glib/gmain.c	2018-01-08 20:00:49.000000000 +0000
+++ glib-2.54.3/glib/gmain.c	2019-04-07 04:12:31.815766012 +0100
@@ -35,11 +35,11 @@
 
 /* Uncomment the next line (and the corresponding line in gpoll.c) to
  * enable debugging printouts if the environment variable
- * G_MAIN_POLL_DEBUG is set to some value.
+ * _G_MAIN_POLL_DEBUG is set to some value.
  */
 /* #define G_MAIN_POLL_DEBUG */
 
-#ifdef _WIN32
+#if defined(_WIN32) || defined(__SWITCH__)
 /* Always enable debugging printout on Windows, as it is more often
  * needed there...
  */
@@ -5026,6 +5026,7 @@
 static void
 dispatch_unix_signals_unlocked (void)
 {
+  #ifndef SWITCH
   gboolean pending[NSIG];
   GSList *node;
   gint i;
@@ -5111,7 +5112,7 @@
             }
         }
     }
-
+  #endif
 }
 
 static void
@@ -5191,6 +5192,7 @@
 static void
 ref_unix_signal_handler_unlocked (int signum)
 {
+  #ifndef SWITCH
   /* Ensure we have the worker context */
   g_get_worker_context ();
   unix_signal_refcount[signum]++;
@@ -5206,11 +5208,13 @@
 #endif
       sigaction (signum, &action, NULL);
     }
+  #endif
 }
 
 static void
 unref_unix_signal_handler_unlocked (int signum)
 {
+  #ifndef SWITCH
   unix_signal_refcount[signum]--;
   if (unix_signal_refcount[signum] == 0)
     {
@@ -5220,6 +5224,7 @@
       sigemptyset (&action.sa_mask);
       sigaction (signum, &action, NULL);
     }
+  #endif
 }
 
 GSource *
@@ -5342,6 +5347,7 @@
 GSource *
 g_child_watch_source_new (GPid pid)
 {
+#ifndef SWITCH
   GSource *source;
   GChildWatchSource *child_watch_source;
 
@@ -5369,6 +5375,9 @@
 #endif /* G_OS_WIN32 */
 
   return source;
+#else
+  return NULL;
+#endif
 }
 
 /**
@@ -5763,7 +5772,7 @@
   if (g_once_init_enter (&initialised))
     {
       /* mask all signals in the worker thread */
-#ifdef G_OS_UNIX
+#if defined G_OS_UNIX && !defined SWITCH
       sigset_t prev_mask;
       sigset_t all;
 
@@ -5772,7 +5781,7 @@
 #endif
       glib_worker_context = g_main_context_new ();
       g_thread_new ("gmain", glib_worker_main, NULL);
-#ifdef G_OS_UNIX
+#if defined G_OS_UNIX && !defined SWITCH
       pthread_sigmask (SIG_SETMASK, &prev_mask, NULL);
 #endif
       g_once_init_leave (&initialised, TRUE);
diff -ur glib-2.54.3-orig/glib/gpoll.c glib-2.54.3/glib/gpoll.c
--- glib-2.54.3-orig/glib/gpoll.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/glib/gpoll.c	2019-04-07 04:30:29.948980061 +0100
@@ -40,7 +40,7 @@
  */
 /* #define G_MAIN_POLL_DEBUG */
 
-#ifdef _WIN32
+#if defined(_WIN32) || defined(__SWITCH__)
 /* Always enable debugging printout on Windows, as it is more often
  * needed there...
  */
@@ -59,7 +59,7 @@
 /* The poll() emulation on OS/X doesn't handle fds=NULL, nfds=0,
  * so we prefer our own poll emulation.
  */
-#if defined(_POLL_EMUL_H_) || defined(BROKEN_POLL)
+#if defined(_POLL_EMUL_H_) || defined(BROKEN_POLL) || defined(__SWITCH__)
 #undef HAVE_POLL
 #endif
 
@@ -74,9 +74,14 @@
 #include <windows.h>
 #endif /* G_OS_WIN32 */
 
+#ifdef __SWITCH__
+#include <switch.h>
+#include <stdio.h>
+#endif
+
 #include "gpoll.h"
 
-#ifdef G_OS_WIN32
+#if defined(G_OS_WIN32) || defined(__SWITCH__)
 #include "gprintf.h"
 #endif
 
@@ -84,6 +89,138 @@
 extern gboolean _g_main_poll_debug;
 #endif
 
+#ifdef __SWITCH__
+static int
+poll_rest (UEvent  **handles,
+           GPollFD *handle_to_fd[],
+           gint     nhandles,
+           gint     timeout)
+{
+  int ready;
+  GPollFD *f;
+  int recursed_result;
+  Waiter waiters[MAX_WAIT_OBJECTS];
+  Result res;
+
+  for(int i = 0; i < nhandles; i++)
+  {
+    waiters[i] = waiterForUEvent(handles[i]);
+  }
+
+  if (_g_main_poll_debug)
+    g_print ("  waitObjects(%d, %d)\n", nhandles, timeout);
+
+    res = waitObjects (&ready, waiters, nhandles, timeout * 1000);
+    if (R_FAILED(res) && res != 0xea01)
+    {
+      gchar *emsg = "idk lol, fucked it";
+      g_warning ("waitObjects failed: %s %08x", emsg, res);
+      g_free (emsg);
+    }
+
+  if (_g_main_poll_debug)
+    g_print ("  wait returns %d\n",
+       ready);
+
+  if (R_FAILED(res) && res != 0xea01)
+    return -1;
+  else if (res == 0xea01)
+    return 0;
+
+  f = handle_to_fd[ready];
+  f->revents = f->events;
+  if (_g_main_poll_debug)
+    g_print ("  got event %p\n", (UEvent*) f->fd);
+
+  /* If no timeout and polling several handles, recurse to poll
+   * the rest of them.
+   */
+  if (timeout == 0 && nhandles > 1)
+  {
+    /* Poll the handles with index > ready */
+          UEvent  **shorter_handles;
+          GPollFD **shorter_handle_to_fd;
+          gint     shorter_nhandles;
+
+          shorter_handles = &handles[ready + 1];
+          shorter_handle_to_fd = &handle_to_fd[ready + 1];
+          shorter_nhandles = nhandles - (ready + 1);
+
+    recursed_result = poll_rest (shorter_handles, shorter_handle_to_fd, shorter_nhandles, 0);
+    return (recursed_result == -1) ? -1 : 1 + recursed_result;
+  }
+  return 1;
+}
+
+gint
+g_poll (GPollFD *fds,
+  guint    nfds,
+  gint     timeout)
+{
+  UEvent* handles[MAX_WAIT_OBJECTS];
+  GPollFD *handle_to_fd[MAX_WAIT_OBJECTS];
+  GPollFD *f;
+  gint nhandles = 0;
+  int retval;
+
+  if (_g_main_poll_debug)
+    g_print ("g_poll: waiting for");
+
+  for (f = fds; f < &fds[nfds]; ++f)
+    {
+      if (f->fd > 0)
+        {
+          if (nhandles == MAX_WAIT_OBJECTS)
+            {
+              g_warning ("Too many handles to wait for!\n");
+              break;
+            }
+          else
+            {
+              if (_g_main_poll_debug)
+                g_print (" %p", (UEvent*) f->fd);
+              handle_to_fd[nhandles] = f;
+              handles[nhandles++] = (UEvent*) f->fd;
+            }
+        }
+      f->revents = 0;
+    }
+
+  if (_g_main_poll_debug)
+    g_print ("\n");
+
+  /* Polling for several things? */
+  if (nhandles > 1)
+    {
+      /* First check if one or several of them are immediately
+       * available
+       */
+      retval = poll_rest (handles, handle_to_fd, nhandles, 0);
+
+      /* If not, and we have a significant timeout, poll again with
+       * timeout then. Note that this will return indication for only
+       * one event, or only for messages.
+       */
+      if (retval == 0 && (timeout == -1 || timeout > 0))
+        retval = poll_rest (handles, handle_to_fd, nhandles, timeout);
+    }
+  else
+    {
+      /* Just polling for one thing, so no need to check first if
+       * available immediately
+       */
+      retval = poll_rest (handles, handle_to_fd, nhandles, timeout);
+    }
+
+  if (retval == -1)
+    for (f = fds; f < &fds[nfds]; ++f)
+      f->revents = 0;
+
+  return retval;
+}
+
+
+#else
 #ifdef HAVE_POLL
 
 /**
@@ -116,6 +253,8 @@
  *
  * Since: 2.20
  **/
+
+
 gint
 g_poll (GPollFD *fds,
 	guint    nfds,
@@ -398,5 +537,5 @@
 }
 
 #endif /* !G_OS_WIN32 */
-
 #endif	/* !HAVE_POLL */
+#endif /* !__SWITCH__ */
\ No newline at end of file
diff -ur glib-2.54.3-orig/glib/grand.c glib-2.54.3/glib/grand.c
--- glib-2.54.3-orig/glib/grand.c	2017-07-14 00:03:39.000000000 +0100
+++ glib-2.54.3/glib/grand.c	2019-03-26 22:13:54.442263566 +0000
@@ -214,9 +214,16 @@
  * 
  * Returns: the new #GRand
  */
+#ifdef SWITCH
+#include <switch/kernel/svc.h>
+#endif
+
 GRand* 
 g_rand_new (void)
 {
+#ifdef SWITCH
+  u64 sys_seeds[2];
+#endif
   guint32 seed[4];
 #ifdef G_OS_UNIX
   static gboolean dev_urandom_exists = TRUE;
@@ -255,11 +262,20 @@
 
   if (!dev_urandom_exists)
     {  
+      #ifndef SWITCH
       g_get_current_time (&now);
       seed[0] = now.tv_sec;
       seed[1] = now.tv_usec;
       seed[2] = getpid ();
       seed[3] = getppid ();
+      #else
+        svcGetInfo(&sys_seeds[0], 11, 0, 0);
+        svcGetInfo(&sys_seeds[1], 11, 0, 1);
+        seed[0] = sys_seeds[0] & 0xffffffff;
+        seed[1] = (sys_seeds[0] & 0xffffffff00000000) >> 32;
+        seed[2] = sys_seeds[1] & 0xffffffff;
+        seed[3] = (sys_seeds[1] & 0xffffffff00000000) >> 32;
+      #endif
     }
 #else /* G_OS_WIN32 */
   /* rand_s() is only available since Visual Studio 2005 and
diff -ur glib-2.54.3-orig/glib/gslice.c glib-2.54.3/glib/gslice.c
--- glib-2.54.3-orig/glib/gslice.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/glib/gslice.c	2019-03-26 22:13:54.445596904 +0000
@@ -407,7 +407,11 @@
     sys_page_size = system_info.dwPageSize;
   }
 #else
+  #ifdef SWITCH
+  sys_page_size = 0x1000;
+  #else
   sys_page_size = sysconf (_SC_PAGESIZE); /* = sysconf (_SC_PAGE_SIZE); = getpagesize(); */
+  #endif
 #endif
   mem_assert (sys_page_size >= 2 * LARGEALIGNMENT);
   mem_assert ((sys_page_size & (sys_page_size - 1)) == 0);
diff -ur glib-2.54.3-orig/glib/gspawn.c glib-2.54.3/glib/gspawn.c
--- glib-2.54.3-orig/glib/gspawn.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/glib/gspawn.c	2019-03-26 22:13:54.445596904 +0000
@@ -300,6 +300,7 @@
               gint                 *exit_status,
               GError              **error)     
 {
+  #ifndef SWITCH
   gint outpipe = -1;
   gint errpipe = -1;
   GPid pid;
@@ -504,6 +505,10 @@
 
       return TRUE;
     }
+  #else
+    *exit_status = 1;
+    return FALSE;
+  #endif
 }
 
 /**
@@ -1073,7 +1078,7 @@
   gint fd;
   gint res = 0;
   
-#ifdef HAVE_SYS_RESOURCE_H
+#if HAVE_SYS_RESOURCE_H && !defined(SWITCH)
   struct rlimit rl;
 #endif
 
@@ -1115,8 +1120,9 @@
    * rlimit trick */
 
 #endif
-  
-#ifdef HAVE_SYS_RESOURCE_H
+
+#ifndef SWITCH
+#if HAVE_SYS_RESOURCE_H
       
   if (getrlimit(RLIMIT_NOFILE, &rl) == 0 && rl.rlim_max != RLIM_INFINITY)
       open_max = rl.rlim_max;
@@ -1127,7 +1133,7 @@
   for (fd = 0; fd < open_max; fd++)
       if ((res = cb (data, fd)) != 0)
           break;
-
+#endif
   return res;
 }
 #endif
@@ -1347,6 +1353,7 @@
                       gint                 *standard_error,
                       GError              **error)     
 {
+#ifndef SWITCH
   GPid pid = -1;
   gint stdin_pipe[2] = { -1, -1 };
   gint stdout_pipe[2] = { -1, -1 };
@@ -1651,6 +1658,9 @@
   close_and_invalidate (&stderr_pipe[1]);
 
   return FALSE;
+#else
+  return FALSE;
+#endif
 }
 
 /* Based on execvp from GNU C Library */
@@ -1660,6 +1670,7 @@
                 gchar      **argv,
                 gchar      **envp)
 {
+#ifndef SWITCH
   /* Count the arguments.  */
   int argc = 0;
   while (argv[argc])
@@ -1687,6 +1698,7 @@
     
     g_free (new_argv);
   }
+#endif
 }
 
 static gchar*
@@ -1706,6 +1718,7 @@
            gboolean     search_path,
            gboolean     search_path_from_envp)
 {
+#ifndef SWITCH
   if (*file == '\0')
     {
       /* We check the simple case first. */
@@ -1843,6 +1856,9 @@
 
   /* Return the error from the last attempt (probably ENOENT).  */
   return -1;
+#else
+  return -1;
+#endif
 }
 
 /**
diff -ur glib-2.54.3-orig/glib/gstdio.c glib-2.54.3/glib/gstdio.c
--- glib-2.54.3-orig/glib/gstdio.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/glib/gstdio.c	2019-03-26 22:13:54.445596904 +0000
@@ -274,8 +274,10 @@
 
   errno = save_errno;
   return retval;
-#else
+#elif !defined(SWITCH)
   return creat (filename, mode);
+#else
+  return -1;
 #endif
 }
 
@@ -841,6 +843,8 @@
 
   errno = save_errno;
   return retval;
+#elif defined SWITCH 
+  return 0;
 #else
   return utime (filename, utb);
 #endif
diff -ur glib-2.54.3-orig/glib/gstdio.h glib-2.54.3/glib/gstdio.h
--- glib-2.54.3-orig/glib/gstdio.h	2017-07-14 00:03:39.000000000 +0100
+++ glib-2.54.3/glib/gstdio.h	2019-03-26 22:13:54.445596904 +0000
@@ -65,7 +65,7 @@
 #define g_rename  rename
 #define g_mkdir   mkdir
 #define g_stat    stat
-#define g_lstat   lstat
+#define g_lstat   stat
 #define g_remove  remove
 #define g_fopen   fopen
 #define g_freopen freopen
diff -ur glib-2.54.3-orig/glib/gtestutils.c glib-2.54.3/glib/gtestutils.c
--- glib-2.54.3-orig/glib/gtestutils.c	2018-01-08 20:00:49.000000000 +0000
+++ glib-2.54.3/glib/gtestutils.c	2019-03-26 22:13:54.448930242 +0000
@@ -964,7 +964,7 @@
            * tests spawn a *lot* of them.  Avoid spamming system crash
            * collection programs such as systemd-coredump and abrt.
            */
-#ifdef HAVE_SYS_RESOURCE_H
+#if defined HAVE_SYS_RESOURCE_H && !defined SWITCH
           {
             struct rlimit limit = { 0, 0 };
             (void) setrlimit (RLIMIT_CORE, &limit);
@@ -2814,7 +2814,7 @@
 g_test_trap_fork (guint64        usec_timeout,
                   GTestTrapFlags test_trap_flags)
 {
-#ifdef G_OS_UNIX
+#if defined G_OS_UNIX && !defined SWITCH
   int stdout_pipe[2] = { -1, -1 };
   int stderr_pipe[2] = { -1, -1 };
   int errsv;
diff -ur glib-2.54.3-orig/glib/gthread.c glib-2.54.3/glib/gthread.c
--- glib-2.54.3-orig/glib/gthread.c	2017-07-14 00:03:39.000000000 +0100
+++ glib-2.54.3/glib/gthread.c	2019-03-26 22:13:54.448930242 +0000
@@ -1037,7 +1037,7 @@
 
   if (count > 0)
     return count;
-#elif defined(_SC_NPROCESSORS_ONLN)
+#elif defined(_SC_NPROCESSORS_ONLN) && !defined(SWITCH)
   {
     int count;
 
diff -ur glib-2.54.3-orig/glib/gthread-posix.c glib-2.54.3/glib/gthread-posix.c
--- glib-2.54.3-orig/glib/gthread-posix.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/glib/gthread-posix.c	2019-03-26 22:14:34.395653043 +0000
@@ -1156,11 +1156,13 @@
 #ifdef HAVE_PTHREAD_ATTR_SETSTACKSIZE
   if (stack_size)
     {
+#ifndef SWITCH
 #ifdef _SC_THREAD_STACK_MIN
       long min_stack_size = sysconf (_SC_THREAD_STACK_MIN);
       if (min_stack_size >= 0)
         stack_size = MAX (min_stack_size, stack_size);
 #endif /* _SC_THREAD_STACK_MIN */
+#endif
       /* No error check here, because some systems can't do it and
        * we simply don't want threads to fail because of that. */
       pthread_attr_setstacksize (&attr, stack_size);
diff -ur glib-2.54.3-orig/glib/gtimer.c glib-2.54.3/glib/gtimer.c
--- glib-2.54.3-orig/glib/gtimer.c	2017-07-14 00:03:39.000000000 +0100
+++ glib-2.54.3/glib/gtimer.c	2019-03-26 22:13:54.448930242 +0000
@@ -236,6 +236,10 @@
   return total;
 }
 
+#ifdef SWITCH
+#include <switch/kernel/svc.h>
+#endif
+
 /**
  * g_usleep:
  * @microseconds: number of microseconds to pause
@@ -252,6 +256,8 @@
 {
 #ifdef G_OS_WIN32
   Sleep (microseconds / 1000);
+#elif defined SWITCH
+  svcSleepThread(microseconds * 1000);
 #else
   struct timespec request, remaining;
   request.tv_sec = microseconds / G_USEC_PER_SEC;
diff -ur glib-2.54.3-orig/glib/gutils.c glib-2.54.3/glib/gutils.c
--- glib-2.54.3-orig/glib/gutils.c	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/glib/gutils.c	2019-03-26 22:13:54.448930242 +0000
@@ -623,7 +623,7 @@
     {
       static UserDatabaseEntry e;
 
-#ifdef G_OS_UNIX
+#if defined(G_OS_UNIX) && !defined(SWITCH)
       {
         struct passwd *pw = NULL;
         gpointer buffer = NULL;
@@ -2431,7 +2431,7 @@
    * https://code.google.com/p/android-developer-preview/issues/detail?id=168
    */
   return issetugid ();
-#elif defined(G_OS_UNIX)
+#elif defined(G_OS_UNIX) && !defined(SWITCH)
   uid_t ruid, euid, suid; /* Real, effective and saved user ID's */
   gid_t rgid, egid, sgid; /* Real, effective and saved group ID's */
 
diff -ur glib-2.54.3-orig/glib/gwakeup.c glib-2.54.3/glib/gwakeup.c
--- glib-2.54.3-orig/glib/gwakeup.c	2017-07-14 00:03:39.000000000 +0100
+++ glib-2.54.3/glib/gwakeup.c	2019-04-04 21:36:11.097833532 +0100
@@ -108,7 +108,47 @@
 }
 
 #else
+#ifdef SWITCH
+#include <stdlib.h>
+#include <stdio.h>
+#include <switch.h>
+#include "glib-unix.h"
+
+GWakeup * g_wakeup_new (void)
+{
+  UEvent *event = malloc(sizeof(struct UEvent));
+  ueventCreate(event, false);
+  return (GWakeup *)event;
+}
+
+void
+g_wakeup_get_pollfd (GWakeup *wakeup,
+                     GPollFD *poll_fd)
+{
+  printf("oh fuck it's all over, pollfd\n");
+  poll_fd->fd = (gintptr) wakeup;
+  poll_fd->events = G_IO_IN;
+}
 
+void
+g_wakeup_acknowledge (GWakeup *wakeup)
+{
+  ueventClear((UEvent*)wakeup);
+}
+
+void
+g_wakeup_signal (GWakeup *wakeup)
+{
+  ueventSignal((UEvent*)wakeup);
+}
+
+void
+g_wakeup_free (GWakeup *wakeup)
+{
+  free((UEvent*)wakeup);
+}
+
+#else
 #include "glib-unix.h"
 #include <fcntl.h>
 
@@ -271,4 +311,5 @@
   g_slice_free (GWakeup, wakeup);
 }
 
-#endif /* !_WIN32 */
+#endif /* switch */
+#endif /* !_WIN32 */
\ No newline at end of file
diff -ur glib-2.54.3-orig/glib/Makefile.am glib-2.54.3/glib/Makefile.am
--- glib-2.54.3-orig/glib/Makefile.am	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/glib/Makefile.am	2019-03-26 22:13:54.448930242 +0000
@@ -425,7 +425,7 @@
 
 if OS_UNIX
 
-INSTALL_PROGS	+= gtester
+#INSTALL_PROGS	+= gtester
 gtester_SOURCES	 = gtester.c
 gtester_LDADD	 = libglib-2.0.la 
 
diff -ur glib-2.54.3-orig/gobject/Makefile.am glib-2.54.3/gobject/Makefile.am
--- glib-2.54.3-orig/gobject/Makefile.am	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/gobject/Makefile.am	2019-03-26 22:13:54.448930242 +0000
@@ -186,7 +186,8 @@
 #
 # programs to compile and install
 #
-bin_PROGRAMS = gobject-query
+#bin_PROGRAMS = gobject-query
+bin_PROGRAMS = 
 bin_SCRIPTS = glib-mkenums glib-genmarshal
 # source files
 gobject_query_SOURCES = gobject-query.c
diff -ur glib-2.54.3-orig/tests/gobject/Makefile.am glib-2.54.3/tests/gobject/Makefile.am
--- glib-2.54.3-orig/tests/gobject/Makefile.am	2018-01-08 20:00:42.000000000 +0000
+++ glib-2.54.3/tests/gobject/Makefile.am	2019-03-26 22:13:54.448930242 +0000
@@ -25,14 +25,14 @@
 	$(NULL)
 
 # Don't install these ones, and keep them out of 'make check' because they take too long...
-noinst_PROGRAMS += \
-	performance				\
-	performance-threaded			\
-	$(NULL)
+#noinst_PROGRAMS += \
+#	performance				\
+#	performance-threaded			\
+#	$(NULL)
 
 # Run the 'installed' tests manually in-tree.
 # This will cause them to be built even if installed tests are disabled.
-check_PROGRAMS += $(installed_test_programs) $(installed_test_extra_programs)
+#check_PROGRAMS += $(installed_test_programs) $(installed_test_extra_programs)
 TESTS += $(installed_test_programs)
 AM_TESTS_ENVIRONMENT = \
 	LIBCHARSET_ALIAS_DIR=$(top_builddir)/glib/libcharset    \
diff -ur glib-2.54.3-orig/tests/Makefile.am glib-2.54.3/tests/Makefile.am
--- glib-2.54.3-orig/tests/Makefile.am	2017-07-14 00:03:39.000000000 +0100
+++ glib-2.54.3/tests/Makefile.am	2019-03-26 22:13:54.452263581 +0000
@@ -8,15 +8,15 @@
 AM_CFLAGS = -g
 
 # Some random programs that appear not to be testcases and not used from any testcases
-check_PROGRAMS += \
-	testgdateparser				\
-	unicode-normalize			\
-	$(NULL)
+#check_PROGRAMS += \
+#	testgdateparser				\
+#	unicode-normalize			\
+#	$(NULL)
 
 # So far, this is the only gtester-ified testcase here
-test_programs = \
-	testglib				\
-	$(NULL)
+#test_programs = \
+#	testglib				\
+#	$(NULL)
 
 # These are not yet gtester-ified, so mark them as for 'installed' only so we
 # don't run them under the framework.  We will handle them manually below.
@@ -55,10 +55,10 @@
 	$(NULL)
 
 if OS_UNIX
-test_programs += \
-	spawn-test				\
-	iochannel-test				\
-	$(NULL)
+#test_programs += \
+#	spawn-test				\
+#	iochannel-test				\
+#	$(NULL)
 
 # These don't appear to work installed and we don't want to run them under gtester either...
 dist_uninstalled_test_extra_scripts = \
@@ -76,7 +76,7 @@
 
 # Run the 'installed' tests manually in-tree.
 # This will cause them to be built even if installed tests are disabled.
-check_PROGRAMS += $(installed_test_programs) $(installed_test_extra_programs)
+#check_PROGRAMS += $(installed_test_programs) $(installed_test_extra_programs)
 check_SCRIPTS += $(installed_test_scripts)
 TESTS += $(installed_test_programs) $(dist_uninstalled_test_extra_scripts)
 AM_TESTS_ENVIRONMENT = \
@@ -86,7 +86,7 @@
 	MALLOC_CHECK_=2						\
 	MALLOC_PERTURB_=$$(($${RANDOM:-256} % 256))
 
-test_programs += module-test
+#test_programs += module-test
 module_test_DEPENDENCIES = libmoduletestplugin_a.la libmoduletestplugin_b.la
 module_test_LDADD = $(top_builddir)/gmodule/libgmodule-2.0.la $(LDADD)
 module_test_LDFLAGS = $(G_MODULE_LDFLAGS)
@@ -122,12 +122,12 @@
 	done
 
 if HAVE_CXX
-installed_test_programs += cxx-test
+#installed_test_programs += cxx-test
 cxx_test_SOURCES = cxx-test.C
 endif
 
 if ENABLE_TIMELOOP
-installed_test_programs += timeloop
+#installed_test_programs += timeloop
 endif
 
 if PLATFORM_WIN32
@@ -138,7 +138,7 @@
 module-test.exp: module-test.o
 	$(DLLTOOL) --output-exp module-test.exp module-test.o
 
-installed_test_extra_programs = spawn-test-win32-gui
+#installed_test_extra_programs = spawn-test-win32-gui
 
 spawn_test_win32_gui_LDFLAGS = -mwindows
 
